emmarin/cl/cl/src/crust/note.rs

```diff
@@ -0,0 +1,135 @@
+use crate::crust::{balance::Unit, nullifier::NullifierCommitment};
+use crate::mantle::ZoneId;
+use crate::{Digest, Hash};
+use rand::RngCore;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct NoteCommitment(pub [u8; 32]);
+
+impl NoteCommitment {
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        &self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
+pub struct NoteWitness {
+    pub value: u64,
+    pub unit: Unit,
+    pub state: [u8; 32],
+    pub nonce: Nonce,
+    pub zone_id: ZoneId,
+}
+
+impl NoteWitness {
+    pub fn new(value: u64, unit: Unit, state: [u8; 32], nonce: Nonce, zone_id: ZoneId) -> Self {
+        Self {
+            value,
+            unit,
+            state,
+            nonce,
+            zone_id,
+        }
+    }
+
+    pub fn stateless(value: u64, unit: Unit, zone_id: ZoneId, rng: impl RngCore) -> Self {
+        Self::new(value, unit, [0u8; 32], Nonce::random(rng), zone_id)
+    }
+
+    pub fn commit(&self, nf_pk: NullifierCommitment) -> NoteCommitment {
+        let mut hasher = Hash::new();
+        hasher.update(b"NOMOS_CL_NOTE_CM");
+        hasher.update(&self.zone_id);
+
+        // COMMIT TO BALANCE
+        hasher.update(self.value.to_le_bytes());
+        hasher.update(self.unit);
+        // Important! we don't commit to the balance blinding factor as that may make the notes linkable.
+
+        // COMMIT TO STATE
+        hasher.update(self.state);
+
+        // COMMIT TO NONCE
+        hasher.update(self.nonce.as_bytes());
+
+        // COMMIT TO NULLIFIER
+        hasher.update(nf_pk.as_bytes());
+
+        let commit_bytes: [u8; 32] = hasher.finalize().into();
+        NoteCommitment(commit_bytes)
+    }
```

Comment:
<img width="538" alt="Screenshot 2025-02-12 at 8 30 08 PM" src="https://github.com/user-attachments/assets/2b89e958-0854-4f1b-9b04-1888ab298cdb" />
This order is not following the spec